File: data-transformer/src/models/equipment.ts
import { Exclude, Expose, Transform, Type } from "class-transformer";
import { DamageRoll } from "./secondary/damage";
import { assignHeaderIds, cleanupHTML, stepHeaders } from "./utils/html";
import type { Publication } from "./secondary/publication";
import { mapToRule } from "./secondary/rule";

export class Equipment {
    @Expose({ name: "_id" }) fvttId!: string;
    name!: string;
    type!: string;

    @Transform(({ obj }) => obj.system.acBonus)
    @Expose()
    acBonus?: number;

    @Transform(({ obj }) => {
        return obj.system.apex ? obj.system.apex.attribute : undefined
    })
    @Expose()
    apex?: string; // attribute

    @Transform(({ obj }) => obj.system.baseItem)
    @Expose()
    baseItem?: string;

    @Transform(({ obj }) => {
        return obj.system.bonus ? obj.system.bonus.value : undefined
    })
    @Expose()
    bonus?: number;

    @Transform(({ obj }) => {
        type Bulk = {
            capacity: number;
            heldOrStowed: number;
            ignored: number;
            value: number;
        };
        const bulkObj = obj.system.bulk as Bulk
        return bulkObj
    })
    @Expose()
    bulk?: any;

    @Transform(({ obj }) => obj.system.category)
    @Expose()
    category?: string;

    @Transform(({ obj }) => obj.system.checkPenalty)
    @Expose()
    checkPenalty?: number;

    @Transform(({ obj }) => obj.system.damage)
    @Type(() => DamageRoll)
    @Expose()
    damage?: DamageRoll;

    @Transform(({ obj }) => {
        let description = obj.system.description.value;
        description = stepHeaders(description);
        description = assignHeaderIds(description);
        description = cleanupHTML(description);
        return description;
    })
    @Expose()
    description!: string;

    @Transform(({ obj }) => obj.system.dexCap)
    @Expose()
    dexCap?: number;

    @Transform(({ obj }) => {
        return obj.system.equipped ? obj.system.equipped.carryType : undefined
    })
    @Expose()
    carryType?: string;

    @Transform(({ obj }) => obj.system.group)
    @Expose()
    group?: string;

    // @Transform(({ obj }) => obj.system.items)
    // @Expose({ name: "items" })
    // @Type(() => Feat)
    // features!: Feat[];

    @Transform(({ obj }) => {
        return obj.system.level ? obj.system.level.value : undefined
    })
    @Expose()
    level?: number;

    @Transform(({ obj }) => {
        let priceObj = obj.system.price;
        return {
            per: priceObj.per,
            cp: priceObj.value.cp,
            gp: priceObj.value.gp,
            sp: priceObj.value.sp
        }
    })
    @Expose()
    price?: any;

    @Transform(({ obj }) => obj.system.publication)
    @Expose()
    publication!: Publication;

    @Transform(({ obj }) => obj.system.quantity)
    @Expose()
    quantity?: number;

    @Transform(({ obj }) => {
        if (obj.system.range && typeof (obj.system.range) !== 'number') {
            return undefined
        }
        return obj.system.range
    })
    @Expose()
    range?: number;

    @Transform(({ obj }) => obj.system.rules.map((v: any) => mapToRule(v)))
    @Expose()
    rules!: any[];

    @Transform(({ obj }) => {
        type Runes = {
            potency?: number;
            property?: string[];
            reinforcing?: number;
            resilient?: number;
            striking?: number;
        };
        const runesObj = obj.system.runes as Runes
        return runesObj
    })
    @Expose()
    runes?: any;

    @Transform(({ obj }) => obj.system.size)
    @Expose()
    size?: string;

    @Transform(({ obj }) => obj.system.slug)
    @Expose()
    slug!: string;

    // TODO: specific

    @Transform(({ obj }) => obj.system.speedPenalty)
    @Expose()
    speedPenalty?: number;

    // TODO: spell

    @Transform(({ obj }) => {
        return obj.system.splashDamage && typeof (obj.system.splashDamage) === 'number' ? obj.system.splashDamage.value : undefined
    })
    @Expose()
    splashDamage?: number;

    @Transform(({ obj }) => obj.system.stackGroup)
    @Expose()
    stackGroup?: string;

    @Transform(({ obj }) => obj.system.stowing)
    @Expose()
    stowing?: boolean;

    @Transform(({ obj }) => obj.system.strength)
    @Expose()
    strength?: number;

    @Transform(({ obj }) => obj.system.traits.rarity)
    @Expose()
    rarity!: string;

    @Transform(({ obj }) => obj.system.traits.value)
    @Expose()
    traits!: string[];

    @Transform(({ obj }) => {
        type Usage = {
            canBeAmmo?: boolean;
            value: string;
        };
        const usageObj = obj.system.usage as Usage
        return usageObj
    })
    @Expose()
    usage?: any;

    @Transform(({ obj }) => {
        type Uses = {
            autoDestroy?: boolean;
            max: number;
            value: number;
        };
        const usesObj = obj.system.uses as Uses
        return usesObj
    })
    @Expose()
    uses?: any;

    @Exclude() effects!: null;
    @Exclude() _stats!: null;
    @Exclude() img!: null;
    @Exclude() system!: null;
}

File: backend/src/models/primary/equipment.rs
use std::collections::{BTreeMap, HashMap};

use crate::helpers::{Either, null_to_default};
use crate::models::{
    Attribute, BoostOrFlaw, DamageRoll, LinkPreview, Publication, Rule, Skill, Summary, SummaryData,
};
use crate::storeable::Storeable;
use bson::oid::ObjectId;
use serde::{Deserialize, Serialize};
use specta::Type;

use super::Feat;

#[derive(Serialize, Deserialize, Debug, Type)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct Equipment {
    #[serde(rename = "_id")]
    pub mongo_id: ObjectId,
    pub fvtt_id: String,
    pub name: String,

    pub ac_bonus: Option<u8>,
    pub apex: Option<Attribute>,
    pub base_item: Option<String>,
    pub bonus: Option<u8>,
    pub bulk: Option<EquipmentBulk>,
    pub category: Option<String>,
    pub check_penalty: Option<i8>,
    pub damage: Option<DamageRoll>,
    pub description: String,
    pub dex_cap: Option<u8>,
    pub carry_type: Option<String>,
    pub group: Option<String>,
    pub level: Option<u8>,
    pub price: Option<EquipmentPrice>,
    pub quantity: Option<u8>,
    pub range: Option<u32>,
    pub runes: Option<EquipmentRunes>,
    pub size: Option<String>,
    pub speed_penalty: Option<i8>,
    pub splash_damage: Option<u8>,
    pub stack_group: Option<String>,
    pub stowing: Option<bool>,
    pub strength: Option<u8>,
    pub usage: Option<EquipmentUsage>,
    pub uses: Option<EquipmentUses>,

    pub rules: Vec<Rule>,
    pub publication: Publication,
    pub rarity: String,
    #[serde(default, deserialize_with = "null_to_default")]
    pub traits: Vec<String>,
    pub slug: String,
    #[serde(rename = "type")]
    pub data_type: String,
}

#[derive(Serialize, Deserialize, Debug, Type)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct EquipmentUses {
    auto_destroy: Option<bool>,
    max: u32,
    value: u32,
}

#[derive(Serialize, Deserialize, Debug, Type)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct EquipmentUsage {
    can_be_ammo: Option<bool>,
    value: String,
}

#[derive(Serialize, Deserialize, Debug, Type)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct EquipmentRunes {
    potency: Option<u8>,
    #[serde(default, deserialize_with = "null_to_default")]
    property: Vec<String>,
    reinforcing: Option<u8>,
    resilient: Option<u8>,
    striking: Option<u8>,
}

#[derive(Serialize, Deserialize, Debug, Type)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct EquipmentPrice {
    per: Option<u8>,
    cp: Option<u32>,
    sp: Option<u32>,
    gp: Option<u32>,
    pp: Option<u32>,
}

#[derive(Serialize, Deserialize, Debug, Type)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct EquipmentBulk {
    held_or_stowed: Option<Either<bool, f32>>,
    capacity: Option<u32>,
    ignored: Option<u32>,
    value: f32,
}

impl Storeable for Equipment {
    fn table_name() -> &'static str {
        "equipment"
    }
}

impl From<Equipment> for Summary {
    fn from(value: Equipment) -> Self {
        let data = vec![SummaryData::String {
            value: value.name.clone(),
            link: None,
            tooltip: None,
            abbreviation: None,
        }];

        Self {
            id: value.fvtt_id,
            name: value.name,
            slug: value.slug,
            data,
        }
    }
}

impl From<Equipment> for LinkPreview {
    fn from(value: Equipment) -> Self {
        Self {
            slug: value.slug,
            name: value.name,
            description: value.description,
            rarity: Some(value.rarity),
            traits: value.traits,
        }
    }
}
